Updated to 0.4.33--receiving error: "Too many open files (os error 24)"

### Problem

Hello!

I just got the notice of the `v0.4.33` update so I updated on my local machine.

When I went to run the local instance `mdbook clean && mdbook serve`, I received an error `mdbook clean && mdbook serve`.

Here is a link to the file in question: [/Users/jlouthan/Projects/Books/theologicus/src/confession-1689/1689-02-3-the-triune-God.md](https://github.com/joelouthan/Theologicus/blob/994294185710e24fc04f365728afe279b5d5e72c/src/confession-1689/1689-02-3-the-triune-God.md)

Note: This is the link to the previous commit. I did see some footnote issues in that particular chapter and cleaned up some links. But v0.4.33 still gave the same error after I cleaned up that particular file.

### Steps

1. Updated `mdbook` to `v0.4.33`
2. Ran local instance: `mdbook clean && mdbook serve`
3. Receive error:
```
2023-08-04 08:26:04 [INFO] (mdbook::book): Book building has started
Warning: The mdbook-toc preprocessor was built against version 0.4.32 of mdbook, but we're being called from version 0.4.33
2023-08-04 08:26:04 [INFO] (mdbook::book): Running the html backend
2023-08-04 08:26:10 [WARN] (mdbook::renderer::html_handlebars::search): searchindex.json is very large (16546566 bytes)
2023-08-04 08:26:10 [INFO] (mdbook::cmd::serve): Serving on: http://localhost:3000
2023-08-04 08:26:10 [INFO] (warp::server): Server::run; addr=[::1]:3000
2023-08-04 08:26:10 [INFO] (warp::server): listening on http://[::1]:3000
2023-08-04 08:26:10 [ERROR] (mdbook::cmd::watch): Error while watching "/Users/jlouthan/Projects/Books/theologicus/src":
    Error { kind: Generic("Too many open files (os error 24) about [\"/Users/jlouthan/Projects/Books/theologicus/src/confession-1689/1689-02-3-the-triune-God.md\"]"), paths: [] }
```

### Possible Solution(s)

_No response_

### Notes

_No response_

### Version

```text
0.4.33
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -37,7 +37,7 @@ toml = "0.5.11" # Do not update, see https://github.com/rust-lang/mdBook/issues/
 topological-sort = "0.2.2"
 
 # Watch feature
-notify = { version = "6.0.1", optional = true, features = ["macos_kqueue"] }
+notify = { version = "6.0.1", optional = true }
 notify-debouncer-mini = { version = "0.3.0", optional = true }
 ignore = { version = "0.4.20", optional = true }
 

--- a/src/utils/fs.rs
+++ b/src/utils/fs.rs
@@ -166,7 +166,7 @@ pub fn copy_files_except_ext(
                         .expect("a file should have a file name...")
                 )
             );
-            fs::copy(
+            copy(
                 entry.path(),
                 &to.join(
                     entry
@@ -178,6 +178,62 @@ pub fn copy_files_except_ext(
         }
     }
     Ok(())
+}
+
+/// Copies a file.
+fn copy<P: AsRef<Path>, Q: AsRef<Path>>(from: P, to: Q) -> Result<()> {
+    let from = from.as_ref();
+    let to = to.as_ref();
+    return copy_inner(from, to)
+        .with_context(|| format!("failed to copy `{}` to `{}`", from.display(), to.display()));
+
+    // This is a workaround for an issue with the macOS file watcher.
+    // Rust's `std::fs::copy` function uses `fclonefileat`, which creates
+    // clones on APFS. Unfortunately fs events seem to trigger on both
+    // sides of the clone, and there doesn't seem to be a way to differentiate
+    // which side it is.
+    // https://github.com/notify-rs/notify/issues/465#issuecomment-1657261035
+    // contains more information.
+    //
+    // This is essentially a copy of the simple copy code path in Rust's
+    // standard library.
+    #[cfg(target_os = "macos")]
+    fn copy_inner(from: &Path, to: &Path) -> Result<()> {
+        use std::fs::OpenOptions;
+        use std::os::unix::fs::{OpenOptionsExt, PermissionsExt};
+
+        let mut reader = File::open(from)?;
+        let metadata = reader.metadata()?;
+        if !metadata.is_file() {
+            anyhow::bail!(
+                "expected a file, `{}` appears to be {:?}",
+                from.display(),
+                metadata.file_type()
+            );
+        }
+        let perm = metadata.permissions();
+        let mut writer = OpenOptions::new()
+            .mode(perm.mode())
+            .write(true)
+            .create(true)
+            .truncate(true)
+            .open(to)?;
+        let writer_metadata = writer.metadata()?;
+        if writer_metadata.is_file() {
+            // Set the correct file permissions, in case the file already existed.
+            // Don't set the permissions on already existing non-files like
+            // pipes/FIFOs or device nodes.
+            writer.set_permissions(perm)?;
+        }
+        std::io::copy(&mut reader, &mut writer)?;
+        Ok(())
+    }
+
+    #[cfg(not(target_os = "macos"))]
+    fn copy_inner(from: &Path, to: &Path) -> Result<()> {
+        fs::copy(from, to)?;
+        Ok(())
+    }
 }
 
 pub fn get_404_output_file(input_404: &Option<String>) -> String {

